packages/core/js-client/src/services/builtins.ts

```diff
@@ -246,120 +301,48 @@ export const builtInServices: Record<
       return success(req.args);
     },
 
-    array_length: (req) => {
-      if (req.args.length !== 1) {
-        return error(
-          "array_length accepts exactly one argument, found: " +
-            req.args.length,
-        );
-      } else {
-        assert(Array.isArray(req.args[0]));
-        return success(req.args[0].length);
-      }
-    },
+    array_length: withSchema(z.tuple([z.array(z.unknown())]))(([arr]) => {
+      return success(arr.length);
+    }),
 
-    identity: (req) => {
-      if (req.args.length > 1) {
-        return error(
-          `identity accepts up to 1 arguments, received ${req.args.length} arguments`,
-        );
-      } else {
-        return success(req.args.length === 0 ? {} : req.args[0]);
-      }
-    },
+    identity: withSchema(z.array(jsonSchema).max(1))((args) => {
+      return success(args.length === 0 ? {} : args[0]);
+    }),
 
-    concat: (req) => {
-      const incorrectArgIndices = req.args //
-        .map((x, i): [boolean, number] => {
-          return [Array.isArray(x), i];
-        })
-        .filter(([isArray]) => {
-          return !isArray;
-        })
-        .map(([, index]) => {
-          return index;
-        });
+    concat: withSchema(z.array(z.array(z.unknown())))((args) => {
+      // concat accepts only 'never' type
+      // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+      const arr = args as never[][];
+      return success([].concat(...arr));
+    }),
```

Comment:
can we do?
```suggestion
concat: withSchema(z.array(z.array(jsonSchema)))((args) => {
      return success(args.flat());
    })
```

Reply:
It will add slow runtime check which is completely useless in this scenario

Reply:
what about `.flat()` with typecast and explanation?

Reply:
Nice!